rust/lance-index/src/vector/ivf/storage.rs

```diff
@@ -0,0 +1,221 @@
+// Copyright 2024 Lance Developers.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::sync::Arc;
+
+use arrow_array::{Array, FixedSizeListArray, Float32Array};
+
+use lance_arrow::FixedSizeListArrayExt;
+use lance_core::{Error, Result};
+use lance_file::{reader::FileReader, writer::FileWriter};
+use lance_io::{traits::WriteExt, utils::read_message};
+use lance_table::io::manifest::ManifestDescribing;
+use log::debug;
+use serde::{Deserialize, Serialize};
+use snafu::{location, Location};
+
+use crate::pb::Ivf as PbIvf;
+
+const IVF_METADATA_KEY: &str = "lance:ivf";
+
+#[warn(dead_code)]
+#[derive(Debug, PartialEq)]
+pub struct IvfMetadadta {
+    /// Centroids of the IVF indices. Can be empty.
+    centroids: Arc<FixedSizeListArray>,
+
+    /// Length of each partition.
+    lengths: Vec<u32>,
+
+    /// pre-computed row offset for each partition, do not persist.
+    partition_row_offsets: Vec<usize>,
+}
+
+/// The IVF metadata stored in the Lance Schema
+#[derive(Serialize, Deserialize, Debug)]
+struct IvfMetadataInSchema {
+    // The file position to store the protobuf binary of IVF metadata.
+    pb_position: usize,
+}
+
+impl IvfMetadadta {
+    pub fn empty(dim: usize) -> Self {
+        let values = Float32Array::from(Vec::<f32>::new());
+        Self {
+            centroids: Arc::new(
+                FixedSizeListArray::try_new_from_values(values, dim as i32)
+                    .expect("Creating empty centroids"),
+            ),
+            lengths: vec![],
+            partition_row_offsets: vec![0],
+        }
+    }
+
+    pub async fn load(reader: &FileReader) -> Result<Self> {
+        let schema = reader.schema();
+        let meta_str = schema.metadata.get(IVF_METADATA_KEY).ok_or(Error::Index {
+            message: format!("{} not found during search", IVF_METADATA_KEY),
+            location: location!(),
+        })?;
+        let ivf_metadata: IvfMetadataInSchema =
+            serde_json::from_str(meta_str).map_err(|e| Error::Index {
+                message: format!("Failed to parse IVF metadata: {}", e),
+                location: location!(),
+            })?;
+
+        let pb: PbIvf = read_message(
+            reader.object_reader.as_ref(),
+            ivf_metadata.pb_position as usize,
+        )
+        .await?;
+        Self::try_from(pb)
+    }
+
+    /// Write the IVF metadata to the lance file.
```

Comment:
I'm not sure you can really call it a "lance file" at this point.  Maybe just "Serialize IVF metadata to protobuf"?

Reply:
It depends what you mean by "lance file".  If you mean it is a file described by https://lancedb.github.io/lance/format.html#file-structure then this is not a lance file (and it will not be one with lance v2).  You have (if memory serves):

```
<Index contents> (as protobuf)
A: <Schema> (as protobuf)
Ptr to A
Magic number
```

Why not just store the entire thing as one protobuf message?

Reply:
In a case of IVF_PQ, this file is a lance file with schema

```
// PQ rows
struct {
   pq_code: FixedSizeList<u8>,
   row_id: uint64,
   // Optionally extra fields
}
```

And the file data structure will have:

```
| PQ rows .... |  Serialized(pb::IVF) | Serialized(pb::ProductQuantization) | ... other binaries |  lance_metadata_with_kvs |
```

For IVF_HNSW, it stores in the same lance file with schema

```
// Hnsw Graph Node / Row
struct {
  row_id: uint64,
  neighbors: List<u32>,
}
```

And the file contains

```
| Hnsw rows. ... | Serialized(pb::Ivf) | Serialized(other graph metadata) | ..  | lance_metadata_with_kvs | 
```

With this, we don't really need to optimize the io path for index files, and we can use more encodings from the lance-file/io in the future to improve performance.